src/d3d11/d3d11_swapchain.cpp

```diff
@@ -241,59 +240,80 @@ namespace dxvk {
     // Flush pending rendering commands before
     auto immediateContext = static_cast<D3D11ImmediateContext*>(deviceContext.ptr());
     immediateContext->Flush();
-
-    // Wait for the sync event so that we respect the maximum frame latency
-    uint64_t frameId = ++m_frameId;
-    m_frameLatencySignal->wait(frameId - GetActualFrameLatency());
     
+    // Present this image, then acquire our next image for
+    // a future present.
+    // This means that Present will always present something
+    // instantly when called. Rather than waiting around
+    // to acquire an image.
+
     for (uint32_t i = 0; i < SyncInterval || i < 1; i++) {
       SynchronizePresent();
 
       if (!m_presenter->hasSwapChain())
         return DXGI_STATUS_OCCLUDED;
 
+      // If we don't have an acquired image and we can't
+      // acquire one, that means we've been occluded.
+      if (m_nextImageIndex == UINT32_MAX && !AcquireNextImage())
+        return DXGI_STATUS_OCCLUDED;
+
       // Presentation semaphores and WSI swap chain image
       vk::PresenterInfo info = m_presenter->info();
-      vk::PresenterSync sync = m_presenter->getSyncSemaphores();
-
-      uint32_t imageIndex = 0;
-
-      VkResult status = m_presenter->acquireNextImage(
-        sync.acquire, VK_NULL_HANDLE, imageIndex);
-
-      while (status != VK_SUCCESS && status != VK_SUBOPTIMAL_KHR) {
-        RecreateSwapChain(m_vsync);
-
-        if (!m_presenter->hasSwapChain())
-          return DXGI_STATUS_OCCLUDED;
-        
-        info = m_presenter->info();
-        sync = m_presenter->getSyncSemaphores();
-
-        status = m_presenter->acquireNextImage(
-          sync.acquire, VK_NULL_HANDLE, imageIndex);
-      }
 
       // Resolve back buffer if it is multisampled. We
       // only have to do it only for the first frame.
       m_context->beginRecording(
         m_device->createCommandList());
       
       m_blitter->presentImage(m_context.ptr(),
-        m_imageViews.at(imageIndex), VkRect2D(),
+        m_imageViews.at(m_nextImageIndex), VkRect2D(),
         m_swapImageView, VkRect2D());
 
       if (m_hud != nullptr)
         m_hud->render(m_context, info.format, info.imageExtent);
       
       if (i + 1 >= SyncInterval)
-        m_context->signal(m_frameLatencySignal, frameId);
+        m_context->signal(m_frameLatencySignal, m_frameId);
 
-      SubmitPresent(immediateContext, sync, i);
+      SubmitPresent(immediateContext, m_nextPresenterSync, i);
     }
 
-    SignalFrameLatencyEvent();
-    return S_OK;
+    // Bump frame id
+    ++m_frameId;
+
+    SyncFrameLatency();
+
+    // Acquire the image for the next present.
+    return AcquireNextImage();
+  }
+
+
+  bool D3D11SwapChain::AcquireNextImage() {
+    m_nextImageIndex = UINT32_MAX;
```

Comment:
Can we use some properly named constant instead of `UINT32_MAX`? This is some non-standard windows crap that we're not really using anywhere at least in the d3d11/dxgi code.

Reply:
UINT32_MAX is from stdint, it's not non-standard Windows crap.

But yeah, it can have it's own constant

Reply:
Added `InvalidImageIndex = ~0u`